60_networks.Rmd

```diff
@@ -0,0 +1,855 @@
+
+```{r}
+# norm_counts: clr-transformed count matrix with samples in rows
+rho_shrink_est <- function(norm_counts, ...) {
+  shrunk_cov <- cov.shrink(norm_counts, ...)
+  p <- ncol(norm_counts)
+  J <- matrix(rep(diag(shrunk_cov), p), p)
+  rho <- 2 * shrunk_cov / (J + t(J))
+  (rho + t(rho)) / 2
+}
+```
+
+```{r}
+# Apply the shrinkage proportionality estimator to the clr-transformed counts
+prop_est <- as(rho_shrink_est(t(assay(tse, "clr"))), "matrix")
+```
+
+A threshold of 0.4 is used to sparsify the estimated proportionality matrix.
+
+```{r}
+# Sparsification
+prop_est[abs(prop_est) < 0.4] <- 0
+
+# Compute dissimilarity matrix
+prop_diss <- sqrt(0.5 * (1 - prop_est))
+
+# Dissimilarities between nodes with zero correlation are set to 1
+# (these nodes are unconnected and thus have maximum dissimilarity).
+prop_diss[prop_est == 0] <- 1
+
+# Compute similarity matrix
+prop_sim <- 1 - prop_diss
+
+# Turn into graph
+prop_graph <- adj2igraph(prop_sim)
+```
+
+### SpiecEasi - MB
+
+This third example estimates partial correlations using the MB approach implemented in `SPIEC-EASI`.
+
+```{r}
+# Extract the count matrix, which spiec.easi needs as input
+counts <- t(assay(tse, "counts"))
+```
+
+```{r, eval=FALSE}
+set.seed(13075)
+se_mb_est <- spiec.easi(counts, method = 'mb', nlambda = 20, 
+                        pulsar.params = list(rep.num = 20))
+```
+
+We store the partial correlations corresponding to the StARS-optimal lambda and turn them into an `igraph` object.
+
+```{r}
+# Get optimal matrix with partial correlations
+se_mb_cor <- as.matrix(getOptBeta(se_mb_est))
+se_mb_cor <- as.matrix(symBeta(se_mb_cor))
+rownames(se_mb_cor) <- colnames(se_mb_cor) <- rownames(tse)
+diag(se_mb_cor) <- 1
+
+# Transform into similarity values
+se_mb_diss <- sqrt(0.5 * (1 - se_mb_cor))
+se_mb_diss[se_mb_cor == 0] <- 1
+se_mb_sim <- 1 - se_mb_diss
+
+# Create igraph object
+se_mb_graph <- adj2igraph(se_mb_sim)
+```
+
+The four graph objects can now be plotted side by side using the same layout for all networks.
+
+```{r, fig.height=10, fig.width=10}
+# Node sizes
+vsize <- colMeans(t(assay(tse, "clr"))) + 6
+
+# Use Fruchterman-Reingold (force-directed) layout
+set.seed(13075)
+lay_fr <- layout_with_fr(se_mb_graph)
+
+par(mfrow = c(2,2))
+plot(sparcc_graph03, layout = lay_fr, vertex.size = vsize, 
+     vertex.label = NA, main = "SparCC (thresh 0.3)")
+plot(sparcc_graph04, layout = lay_fr, vertex.size = vsize, 
+     vertex.label = NA, main = "SparCC (thresh 0.4)")
+plot(prop_graph, layout = lay_fr, vertex.size = vsize, 
+     vertex.label = NA, main = "Shrinkage proportionality\n(thresh 0.4)")
+plot(se_mb_graph, layout = lay_fr, vertex.size = vsize, 
+     vertex.label = NA, main = "SpiecEasi (MB)")
+```
```

Comment:
The igraph package (used here) is a state-of-the-art package for visualizing networks in R. We could use the [ggnet2](https://briatte.github.io/ggnet/) package instead but I don't see any advantage because the output would look quite the same.